### ISSUE TITLE
Addressing the Performance Gap: LoRA-Fine-Tuned GRM Evaluation on reward-bench

### ISSUE THREAD

**Issue by GuoweiRong** [YangRui2015/Generalizable-Reward-Model#8]
I fine-tuned GRM using LoRA and loaded the vhead and LoRA weights using the evaluation method from your code, but the performance on reward-bench still falls short of the paper's results. How did you perform the evaluation on reward-bench?Thank you very much!

**Comment by YangRui2015** [comment]
Hi, could you check whether this comment solves your issue?
https://github.com/YangRui2015/Generalizable-Reward-Model/issues/4#issuecomment-2781207332

**Comment by GuoweiRong** [comment]
Yes, I saw your comment. While the pipeline can handle full fine-tuning cases immediately, using LoRA requires reinitializing the model.Following your evaluation code:fine-tuned models load a head, whereas non-fine-tuned ones use GRewardModel to load the model. I rewrote the bench-function and successfully loaded the vhead using your model-loading code, but the performance remains mediocre.



---- Replied Message ----
| From | ***@***.***> |
| Date | 06/06/2025 01:17 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [YangRui2015/Generalizable-Reward-Model] Addressing the Performance Gap: LoRA-Fine-Tuned GRM Evaluation on reward-bench (Issue #8) |
YangRui2015 left a comment (YangRui2015/Generalizable-Reward-Model#8)

Hi, could you check whether this comment solves your issue?
#4 (comment)

**Comment by YangRui2015** [comment]
Since the experiments were conducted quite a while ago, I recall that I first merged the LoRA model and then used it just like a fully trained reward model. The following code shows how the LoRA merge was done.


```python
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union
import evaluate
import numpy as np
import os
import torch
import torch.nn as nn
from datasets import load_dataset
from peft import LoraConfig, TaskType, get_peft_model, PeftModel
from transformers import (
   AutoModelForSequenceClassification,
   AutoTokenizer,
   HfArgumentParser,
   PreTrainedTokenizerBase,
   Trainer,
   TrainerCallback,
   TrainingArguments,
   LlamaTokenizer,
)
from transformers.utils import PaddingStrategy


# Define and parse arguments.
@dataclass
class ScriptArguments:
   base_model: Optional[str] =  field(default="google/gemma-2b-it")
   peft_name: Optional[str] = field(default='eward_models_unified_new/gemma-2b-it_gemma2b-it_linear_sftreg0.005_lora64_len4096_mixdata_5e-06/logs/checkpoint-8582')
   save_dir: Optional[str] = field(default='./reward_models_unified_merge/gemma2b-it_linear_sftreg0.005_lora64_len4096_mixdata_5e-06')

parser = HfArgumentParser(ScriptArguments)
script_args = parser.parse_args_into_dataclasses()[0]

tokenizer = AutoTokenizer.from_pretrained(script_args.base_model, use_fast = False)
if 'Llama' in script_args.base_model or 'addpadtoken' in script_args.peft_name:
   tokenizer.add_special_tokens({'pad_token': '[PAD]'})
else:
   tokenizer.pad_token = tokenizer.eos_token

model = AutoModelForSequenceClassification.from_pretrained(
   script_args.base_model, num_labels=1, device_map='cpu',
   torch_dtype=torch.bfloat16,
)
model.resize_token_embeddings(len(tokenizer))
model.config.pad_token_id = tokenizer.pad_token_id
model = PeftModel.from_pretrained(model, script_args.peft_name)
model = model.merge_and_unload()

print("Saving the merged model")
model.save_pretrained(script_args.save_dir)
tokenizer.save_pretrained(script_args.save_dir)
```

**Comment by GuoweiRong** [comment]
Ok, thank you very much for your response. I generally understand the framework you outlined, and my approach aligns with yours — I directly used your provided method for loading the model with the vhead. I still have a few questions:
After using your vhead-enabled model loader, should I simply register the GRM details in reward-bench as mentioned in that comment and run the evaluation via the benchmark's CLI mode? Currently, I've rewritten an `evaluate.py` based on the benchmark's theory, reimplemented the pipeline, and used the forward function from `gutil` in your `rm_eval` code. Is this correct? Or can I directly use the approach described in your comment?  
Thanks again for your patient guidance.



---- Replied Message ----
| From | ***@***.***> |
| Date | 06/06/2025 03:03 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [YangRui2015/Generalizable-Reward-Model] Addressing the Performance Gap: LoRA-Fine-Tuned GRM Evaluation on reward-bench (Issue #8) |
YangRui2015 left a comment (YangRui2015/Generalizable-Reward-Model#8)

Since the experiments were conducted quite a while ago, I recall that I first merged the LoRA model and then used it just like a fully trained reward model. The following code shows how the LoRA merge was done.

fromdataclassesimportdataclass, fieldfromtypingimportAny, Dict, List, Optional, Unionimportevaluateimportnumpyasnpimportosimporttorchimporttorch.nnasnnfromdatasetsimportload_datasetfrompeftimportLoraConfig, TaskType, get_peft_model, PeftModelfromtransformersimport (
   AutoModelForSequenceClassification,
   AutoTokenizer,
   HfArgumentParser,
   PreTrainedTokenizerBase,
   Trainer,
   TrainerCallback,
   TrainingArguments,
   LlamaTokenizer,
)
fromtransformers.utilsimportPaddingStrategy# Define and parse ***@***.***:
   base_model: Optional[str] =field(default="google/gemma-2b-it")
   peft_name: Optional[str] =field(default='eward_models_unified_new/gemma-2b-it_gemma2b-it_linear_sftreg0.005_lora64_len4096_mixdata_5e-06/logs/checkpoint-8582')
   save_dir: Optional[str] =field(default='./reward_models_unified_merge/gemma2b-it_linear_sftreg0.005_lora64_len4096_mixdata_5e-06')

parser=HfArgumentParser(ScriptArguments)
script_args=parser.parse_args_into_dataclasses()[0]

tokenizer=AutoTokenizer.from_pretrained(script_args.base_model, use_fast=False)
if'Llama'inscript_args.base_modelor'addpadtoken'inscript_args.peft_name:
   tokenizer.add_special_tokens({'pad_token': '[PAD]'})
else:
   tokenizer.pad_token=tokenizer.eos_tokenmodel=AutoModelForSequenceClassification.from_pretrained(
   script_args.base_model, num_labels=1, device_map='cpu',
   torch_dtype=torch.bfloat16,
)
model.resize_token_embeddings(len(tokenizer))
model.config.pad_token_id=tokenizer.pad_token_idmodel=PeftModel.from_pretrained(model, script_args.peft_name)
model=model.merge_and_unload()

print("Saving the merged model")
model.save_pretrained(script_args.save_dir)
tokenizer.save_pretrained(script_args.save_dir)

**Comment by YangRui2015** [comment]
I recommend registering the model and using the benchmark's CLI command to run it—that's what I did previously. You can also modify their original code to load GRM with lora in their pipeline.

**Comment by 1sir3** [comment]
明白，衷心感谢您的回复



---- Replied Message ----
| From | ***@***.***> |
| Date | 06/07/2025 07:02 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [YangRui2015/Generalizable-Reward-Model] Addressing the Performance Gap: LoRA-Fine-Tuned GRM Evaluation on reward-bench (Issue #8) |
YangRui2015 left a comment (YangRui2015/Generalizable-Reward-Model#8)

I recommend registering the model and using the benchmark's CLI command to run it—that's what I did previously. You can also modify their original code to load GRM with lora in their pipeline.

**Comment by GuoweiRong** [comment]
Got it. Sincere appreciation for your guidance.






---- Replied Message ----
| From | ***@***.***> |
| Date | 06/07/2025 07:02 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [YangRui2015/Generalizable-Reward-Model] Addressing the Performance Gap: LoRA-Fine-Tuned GRM Evaluation on reward-bench (Issue #8) |
YangRui2015 left a comment (YangRui2015/Generalizable-Reward-Model#8)

I recommend registering the model and using the benchmark's CLI command to run it—that's what I did previously. You can also modify their original code to load GRM with lora in their pipeline.
